{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "waiting-nelson",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'2.4.0'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from tensorflow import keras\n",
    "keras.__version__"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "sought-index",
   "metadata": {},
   "source": [
    "To start using zoo.orca, we need to first initialize orca context. Here we specify local or distributed mode. In this example, we choose the local mode."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "described-timer",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initializing orca context\n",
      "Current pyspark location is : /intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/__init__.py\n",
      "Start to getOrCreate SparkContext\n",
      "pyspark_submit_args is:  --driver-class-path /home/zhenhao/analytics-zoo/zoo/target/analytics-zoo-bigdl_0.12.1-spark_2.4.3-0.10.0-SNAPSHOT-dist-all/lib/analytics-zoo-bigdl_0.12.1-spark_2.4.3-0.10.0-SNAPSHOT-jar-with-dependencies.jar pyspark-shell \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "SLF4J: Class path contains multiple SLF4J bindings.\n",
      "SLF4J: Found binding in [jar:file:/home/zhenhao/analytics-zoo/zoo/target/analytics-zoo-bigdl_0.12.1-spark_2.4.3-0.10.0-SNAPSHOT-dist-all/lib/analytics-zoo-bigdl_0.12.1-spark_2.4.3-0.10.0-SNAPSHOT-jar-with-dependencies.jar!/org/slf4j/impl/StaticLoggerBinder.class]\n",
      "SLF4J: Found binding in [jar:file:/intern/spark/spark-2.4.3-bin-hadoop2.7/jars/slf4j-log4j12-1.7.16.jar!/org/slf4j/impl/StaticLoggerBinder.class]\n",
      "SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.\n",
      "SLF4J: Actual binding is of type [org.slf4j.impl.Log4jLoggerFactory]\n",
      "Setting default log level to \"WARN\".\n",
      "To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2021-02-02 20:16:20 WARN  Utils:66 - Your hostname, intern01 resolves to a loopback address: 127.0.1.1; using 10.239.44.107 instead (on interface eno1)\n",
      "2021-02-02 20:16:20 WARN  Utils:66 - Set SPARK_LOCAL_IP if you need to bind to another address\n",
      "2021-02-02 20:16:21 WARN  NativeCodeLoader:62 - Unable to load native-hadoop library for your platform... using builtin-java classes where applicable\n",
      "cls.getname: com.intel.analytics.bigdl.python.api.Sample\n",
      "BigDLBasePickler registering: bigdl.util.common  Sample\n",
      "cls.getname: com.intel.analytics.bigdl.python.api.EvaluatedResult\n",
      "BigDLBasePickler registering: bigdl.util.common  EvaluatedResult\n",
      "cls.getname: com.intel.analytics.bigdl.python.api.JTensor\n",
      "BigDLBasePickler registering: bigdl.util.common  JTensor\n",
      "cls.getname: com.intel.analytics.bigdl.python.api.JActivity\n",
      "Successfully got a SparkContextBigDLBasePickler registering: bigdl.util.common  JActivity\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "User settings:\n",
      "\n",
      "   KMP_AFFINITY=granularity=fine,compact,1,0\n",
      "   KMP_BLOCKTIME=0\n",
      "   KMP_SETTINGS=1\n",
      "   OMP_NUM_THREADS=1\n",
      "\n",
      "Effective settings:\n",
      "\n",
      "   KMP_ABORT_DELAY=0\n",
      "   KMP_ADAPTIVE_LOCK_PROPS='1,1024'\n",
      "   KMP_ALIGN_ALLOC=64\n",
      "   KMP_ALL_THREADPRIVATE=128\n",
      "   KMP_ATOMIC_MODE=2\n",
      "   KMP_BLOCKTIME=0\n",
      "   KMP_CPUINFO_FILE: value is not defined\n",
      "   KMP_DETERMINISTIC_REDUCTION=false\n",
      "   KMP_DEVICE_THREAD_LIMIT=2147483647\n",
      "   KMP_DISP_HAND_THREAD=false\n",
      "   KMP_DISP_NUM_BUFFERS=7\n",
      "   KMP_DUPLICATE_LIB_OK=false\n",
      "   KMP_FORCE_REDUCTION: value is not defined\n",
      "   KMP_FOREIGN_THREADS_THREADPRIVATE=true\n",
      "   KMP_FORKJOIN_BARRIER='2,2'\n",
      "   KMP_FORKJOIN_BARRIER_PATTERN='hyper,hyper'\n",
      "   KMP_FORKJOIN_FRAMES=true\n",
      "   KMP_FORKJOIN_FRAMES_MODE=3\n",
      "   KMP_GTID_MODE=3\n",
      "   KMP_HANDLE_SIGNALS=false\n",
      "   KMP_HOT_TEAMS_MAX_LEVEL=1\n",
      "   KMP_HOT_TEAMS_MODE=0\n",
      "   KMP_INIT_AT_FORK=true\n",
      "   KMP_INIT_WAIT=2048\n",
      "   KMP_ITT_PREPARE_DELAY=0\n",
      "   KMP_LIBRARY=throughput\n",
      "   KMP_LOCK_KIND=queuing\n",
      "   KMP_MALLOC_POOL_INCR=1M\n",
      "   KMP_NEXT_WAIT=1024\n",
      "   KMP_NUM_LOCKS_IN_BLOCK=1\n",
      "   KMP_PLAIN_BARRIER='2,2'\n",
      "   KMP_PLAIN_BARRIER_PATTERN='hyper,hyper'\n",
      "   KMP_REDUCTION_BARRIER='1,1'\n",
      "   KMP_REDUCTION_BARRIER_PATTERN='hyper,hyper'\n",
      "   KMP_SCHEDULE='static,balanced;guided,iterative'\n",
      "   KMP_SETTINGS=true\n",
      "   KMP_SPIN_BACKOFF_PARAMS='4096,100'\n",
      "   KMP_STACKOFFSET=64\n",
      "   KMP_STACKPAD=0\n",
      "   KMP_STACKSIZE=4M\n",
      "   KMP_STORAGE_MAP=false\n",
      "   KMP_TASKING=2\n",
      "   KMP_TASKLOOP_MIN_TASKS=0\n",
      "   KMP_TASK_STEALING_CONSTRAINT=1\n",
      "   KMP_TEAMS_THREAD_LIMIT=8\n",
      "   KMP_TOPOLOGY_METHOD=all\n",
      "   KMP_USER_LEVEL_MWAIT=false\n",
      "   KMP_VERSION=false\n",
      "   KMP_WARNINGS=true\n",
      "   OMP_AFFINITY_FORMAT='OMP: pid %P tid %T thread %n bound to OS proc set {%a}'\n",
      "   OMP_ALLOCATOR=omp_default_mem_alloc\n",
      "   OMP_CANCELLATION=false\n",
      "   OMP_DEFAULT_DEVICE=0\n",
      "   OMP_DISPLAY_AFFINITY=false\n",
      "   OMP_DISPLAY_ENV=false\n",
      "   OMP_DYNAMIC=false\n",
      "   OMP_MAX_ACTIVE_LEVELS=2147483647\n",
      "   OMP_MAX_TASK_PRIORITY=0\n",
      "   OMP_NESTED=false\n",
      "   OMP_NUM_THREADS='1'\n",
      "   OMP_PLACES: value is not defined\n",
      "   OMP_PROC_BIND='intel'\n",
      "   OMP_SCHEDULE='static'\n",
      "   OMP_STACKSIZE=4M\n",
      "   OMP_TARGET_OFFLOAD=DEFAULT\n",
      "   OMP_THREAD_LIMIT=2147483647\n",
      "   OMP_TOOL=enabled\n",
      "   OMP_TOOL_LIBRARIES: value is not defined\n",
      "   OMP_WAIT_POLICY=PASSIVE\n",
      "   KMP_AFFINITY='noverbose,warnings,respect,granularity=fine,compact,1,0'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from zoo.orca import init_orca_context, stop_orca_context\n",
    "from zoo.orca import OrcaContext\n",
    "\n",
    "OrcaContext.log_output = True # recommended to set it to True when running Analytics Zoo in Jupyter notebook (this will display terminal's  stdout and stderr in the Jupyter notebook).\n",
    "\n",
    "cluster_mode = \"local\"\n",
    "\n",
    "if cluster_mode == \"local\":  \n",
    "    init_orca_context(cluster_mode=\"local\", cores=4) # run in local mode\n",
    "elif cluster_mode == \"k8s\":  \n",
    "    init_orca_context(cluster_mode=\"k8s\", num_nodes=2, cores=2) # run on K8s cluster\n",
    "elif cluster_mode == \"yarn\":  \n",
    "    init_orca_context(cluster_mode=\"yarn-client\", num_nodes=2, cores=2) # run on Hadoop YARN cluster"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "organizational-bible",
   "metadata": {},
   "source": [
    "# Text generation with LSTM\n",
    "This notebook contains the code samples found in Chapter 8, Section 1 of Deep Learning with Python. Note that the original text features far more content, in particular further explanations and figures: in this notebook, you will only find source code and related comments.\n",
    "\n",
    "-----------------------------------------\n",
    "\n",
    "[...]\n",
    "\n",
    "## Implementing character-level LSTM text generation\n",
    "Let's put these ideas in practice in a Keras implementation. The first thing we need is a lot of text data that we can use to learn a language model. You could use any sufficiently large text file or set of text files -- Wikipedia, the Lord of the Rings, etc. In this example we will use some of the writings of Nietzsche, the late-19th century German philosopher (translated to English). The language model we will learn will thus be specifically a model of Nietzsche's writing style and topics of choice, rather than a more generic model of the English language.\n",
    "\n",
    "## Preparing the data\n",
    "Let's start by downloading the corpus and converting it to lowercase:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "lesbian-floating",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Corpus length: 600893\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "path = keras.utils.get_file(\n",
    "    'nietzsche.txt',\n",
    "    origin='https://s3.amazonaws.com/text-datasets/nietzsche.txt')\n",
    "text = open(path).read().lower()\n",
    "print('Corpus length:', len(text))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "metallic-patrol",
   "metadata": {},
   "source": [
    "Next, we will extract partially-overlapping sequences of length maxlen, one-hot encode them and pack them in a 3D Numpy array `x` of shape `(sequences, maxlen, unique_characters)`. Simultaneously, we prepare a array `y` containing the corresponding targets: the one-hot encoded characters that come right after each extracted sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "instrumental-question",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of sequences: 200278\n",
      "Unique characters: 57\n",
      "Vectorization...\n"
     ]
    }
   ],
   "source": [
    "# Length of extracted character sequences\n",
    "maxlen = 60\n",
    "\n",
    "# We sample a new sequence every `step` characters\n",
    "step = 3\n",
    "\n",
    "# This holds our extracted sequences\n",
    "sentences = []\n",
    "\n",
    "# This holds the targets (the follow-up characters)\n",
    "next_chars = []\n",
    "\n",
    "for i in range(0, len(text) - maxlen, step):\n",
    "    sentences.append(text[i: i + maxlen])\n",
    "    next_chars.append(text[i + maxlen])\n",
    "print('Number of sequences:', len(sentences))\n",
    "\n",
    "# List of unique characters in the corpus\n",
    "chars = sorted(list(set(text)))\n",
    "print('Unique characters:', len(chars))\n",
    "# Dictionary mapping unique characters to their index in `chars`\n",
    "char_indices = dict((char, chars.index(char)) for char in chars)\n",
    "\n",
    "# Next, one-hot encode the characters into binary arrays.\n",
    "print('Vectorization...')\n",
    "x = np.zeros((len(sentences), maxlen, len(chars)), dtype=np.bool)\n",
    "y = np.zeros((len(sentences), len(chars)), dtype=np.bool)\n",
    "for i, sentence in enumerate(sentences):\n",
    "    for t, char in enumerate(sentence):\n",
    "        x[i, t, char_indices[char]] = 1\n",
    "    y[i, char_indices[next_chars[i]]] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "beneficial-dealing",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(200278, 60, 57)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "laughing-amateur",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(200278, 57)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "closed-former",
   "metadata": {},
   "source": [
    "\n",
    "## Building the network\n",
    "Our network is a single LSTM layer followed by a Dense classifier and softmax over all possible characters. But let us note that recurrent neural networks are not the only way to do sequence data generation; 1D convnets also have proven extremely successful at it in recent times. Since our targets are one-hot encoded, we will use `categorical_crossentropy` as the loss to train the model:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "absent-spine",
   "metadata": {},
   "outputs": [],
   "source": [
    "from tensorflow.keras import layers\n",
    "\n",
    "def build_model(config):\n",
    "    model = keras.Sequential()\n",
    "    model.add(layers.LSTM(128, input_shape=(maxlen, len(chars))))\n",
    "    model.add(layers.Dense(len(chars), activation='softmax'))\n",
    "    optimizer = keras.optimizers.RMSprop(lr=0.01)\n",
    "    model.compile(loss='categorical_crossentropy', optimizer=optimizer)\n",
    "    return model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "chinese-sigma",
   "metadata": {},
   "source": [
    "We also create a `data_creator` function that help us import the data into the orca estimator."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "comic-operation",
   "metadata": {},
   "outputs": [],
   "source": [
    "import tensorflow as tf\n",
    "\n",
    "def train_data_creator(config):\n",
    "    dataset = tf.data.Dataset.from_tensor_slices((x, y))\n",
    "    dataset = dataset.repeat()\n",
    "    dataset = dataset.shuffle(1000)\n",
    "    dataset = dataset.batch(config[\"batch_size\"])\n",
    "    return dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "smoking-nirvana",
   "metadata": {},
   "source": [
    "Training the language model and sampling from it\n",
    "Given a trained model and a seed text snippet, we generate new text by repeatedly:\n",
    "\n",
    "* 1) Drawing from the model a probability distribution over the next character given the text available so far\n",
    "* 2) Reweighting the distribution to a certain \"temperature\"\n",
    "* 3) Sampling the next character at random according to the reweighted distribution\n",
    "* 4) Adding the new character at the end of the available text\n",
    "\n",
    "This is the code we use to reweight the original probability distribution coming out of the model, and draw a character index from it (the \"sampling function\"):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "accessory-johnson",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sample(preds, temperature=1.0):\n",
    "    preds = np.asarray(preds).astype('float64')\n",
    "    preds = np.log(preds) / temperature\n",
    "    exp_preds = np.exp(preds)\n",
    "    preds = exp_preds / np.sum(exp_preds)\n",
    "    probas = np.random.multinomial(1, preds, 1)\n",
    "    return np.argmax(probas)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "violent-cleaner",
   "metadata": {},
   "source": [
    "Finally, this is the loop where we repeatedly train and generated text. We start generating text using a range of different temperatures after every epoch. This allows us to see how the generated text evolves as the model starts converging, as well as the impact of temperature in the sampling strategy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "encouraging-asbestos",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-02-02 20:16:44,936\tINFO resource_spec.py:212 -- Starting Ray with 17.68 GiB memory available for workers and up to 8.85 GiB for objects. You can adjust these settings with ray.init(memory=<bytes>, object_store_memory=<bytes>).\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'node_ip_address': '10.239.44.107', 'redis_address': '10.239.44.107:23207', 'object_store_address': '/tmp/ray/session_2021-02-02_20-16-44_935881_18041/sockets/plasma_store', 'raylet_socket_name': '/tmp/ray/session_2021-02-02_20-16-44_935881_18041/sockets/raylet', 'webui_url': None, 'session_dir': '/tmp/ray/session_2021-02-02_20-16-44_935881_18041'}\n",
      "\u001b[2m\u001b[36m(pid=18170)\u001b[0m 2021-02-02 20:16:46.007307: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'libcudart.so.10.1'; dlerror: libcudart.so.10.1: cannot open shared object file: No such file or directory\n",
      "\u001b[2m\u001b[36m(pid=18170)\u001b[0m 2021-02-02 20:16:46.007333: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.\n",
      "\u001b[2m\u001b[36m(pid=18170)\u001b[0m 2021-02-02 20:16:46.838793: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory\n",
      "\u001b[2m\u001b[36m(pid=18170)\u001b[0m 2021-02-02 20:16:46.838815: W tensorflow/stream_executor/cuda/cuda_driver.cc:312] failed call to cuInit: UNKNOWN ERROR (303)\n",
      "\u001b[2m\u001b[36m(pid=18170)\u001b[0m 2021-02-02 20:16:46.838830: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (intern01): /proc/driver/nvidia/version does not exist\n",
      "\u001b[2m\u001b[36m(pid=18170)\u001b[0m 2021-02-02 20:16:46.839003: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN)to use the following CPU instructions in performance-critical operations:  AVX2 FMA\n",
      "\u001b[2m\u001b[36m(pid=18170)\u001b[0m To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "\u001b[2m\u001b[36m(pid=18170)\u001b[0m 2021-02-02 20:16:46.860625: I tensorflow/core/platform/profile_utils/cpu_utils.cc:104] CPU Frequency: 3591785000 Hz\n",
      "\u001b[2m\u001b[36m(pid=18170)\u001b[0m 2021-02-02 20:16:46.860946: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x55a150e80ac0 initialized for platform Host (this does not guarantee that XLA will be used). Devices:\n",
      "\u001b[2m\u001b[36m(pid=18170)\u001b[0m 2021-02-02 20:16:46.860977: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version\n",
      "\u001b[2m\u001b[36m(pid=18170)\u001b[0m 2021-02-02 20:16:46.865814: I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:301] Initialize GrpcChannelCache for job worker -> {0 -> 10.239.44.107:44345}\n",
      "\u001b[2m\u001b[36m(pid=18170)\u001b[0m 2021-02-02 20:16:46.866066: I tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc:405] Started server with target: grpc://10.239.44.107:44345\n"
     ]
    }
   ],
   "source": [
    "from zoo.orca.learn.tf2 import Estimator\n",
    "import random\n",
    "import sys\n",
    "\n",
    "batch_size=128\n",
    "est = Estimator.from_keras(model_creator=build_model,\n",
    "                           config={},\n",
    "                           workers_per_node=1,\n",
    "                           verbose=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "developmental-belief",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "epoch 1\n",
      "\u001b[2m\u001b[36m(pid=14333)\u001b[0m WARNING:tensorflow:From /home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/data/ops/multi_device_iterator_ops.py:601: get_next_as_optional (from tensorflow.python.data.ops.iterator_ops) is deprecated and will be removed in a future version.\n",
      "\u001b[2m\u001b[36m(pid=14333)\u001b[0m Instructions for updating:\n",
      "\u001b[2m\u001b[36m(pid=14333)\u001b[0m Use `tf.data.Iterator.get_next_as_optional()` instead.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\r",
      "                                                                                \r",
      "2021-02-02 15:52:50,585\tWARNING worker.py:1072 -- WARNING: 12 PYTHON workers have been started. This could be a result of using a large number of actors, or it could be a consequence of using nested tasks (see https://github.com/ray-project/ray/issues/3644) for some a discussion of workarounds.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- Generating with seed: \" been only too glad to\n",
      "look upon itself as the ultimate end \"\n",
      "------ temperature: 0.2\n",
      " been only too glad to\n",
      "look upon itself as the ultimate end Try to unpersist an uncached rdd\n",
      "Try to unpersist an uncached rdd\n",
      "Try to unpersist an uncached rdd\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-02-02 15:52:50,937\tWARNING worker.py:1072 -- WARNING: 13 PYTHON workers have been started. This could be a result of using a large number of actors, or it could be a consequence of using nested tasks (see https://github.com/ray-project/ray/issues/3644) for some a discussion of workarounds.\n",
      "2021-02-02 15:52:51,201\tWARNING worker.py:792 -- When connecting to an existing cluster, _internal_config must match the cluster's _internal_config.\n",
      "2021-02-02 15:52:51,242\tWARNING worker.py:792 -- When connecting to an existing cluster, _internal_config must match the cluster's _internal_config.\n",
      "2021-02-02 15:52:51,243\tWARNING worker.py:792 -- When connecting to an existing cluster, _internal_config must match the cluster's _internal_config.\n",
      "2021-02-02 15:52:51,247\tWARNING worker.py:792 -- When connecting to an existing cluster, _internal_config must match the cluster's _internal_config.\n",
      "2021-02-02 15:52:51,521\tWARNING worker.py:792 -- When connecting to an existing cluster, _internal_config must match the cluster's _internal_config.\n",
      "2021-02-02 15:52:51,521\tWARNING worker.py:792 -- When connecting to an existing cluster, _internal_config must match the cluster's _internal_config.\n",
      "2021-02-02 15:52:51,521\tWARNING worker.py:792 -- When connecting to an existing cluster, _internal_config must match the cluster's _internal_config.\n",
      "2021-02-02 15:52:51,522\tWARNING worker.py:792 -- When connecting to an existing cluster, _internal_config must match the cluster's _internal_config.\n",
      "1/1 [==============================] - 0s 185us/step\n",
      "1/1 [==============================] - 0s 185us/step\n",
      "1/1 [==============================] - 0s 185us/step\n",
      "[Stage 10:==============>                                           (1 + 3) / 4]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(pid=14333)\u001b[0m \r",
      "1/1 [==============================] - ETA: 0s\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\r",
      "1/1 [==============================] - 0s 185us/step\n",
      "2021-02-02 15:52:52 ERROR Executor:91 - Exception in task 1.0 in stage 10.0 (TID 41)\n",
      "org.apache.spark.api.python.PythonException: Traceback (most recent call last):\n",
      "  File \"/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/worker.py\", line 377, in main\n",
      "    process()\n",
      "  File \"/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/worker.py\", line 372, in process\n",
      "    serializer.dump_stream(func(split_index, iterator), outfile)\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/data/ray_rdd.py\", line 133, in <lambda>\n",
      "    lambda idx, _: get_from_ray(idx, address, password, meta_store_name))\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/data/ray_rdd.py\", line 99, in get_from_ray\n",
      "    partition = ray.get(object_id)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/ray/worker.py\", line 1513, in get\n",
      "    raise value.as_instanceof_cause()\n",
      "ray.exceptions.RayTaskError(OverflowError): \u001b[36mray::TFRunner.predict()\u001b[39m (pid=14333, ip=10.239.44.107)\n",
      "  File \"python/ray/_raylet.pyx\", line 452, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 407, in ray._raylet.execute_task.function_executor\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in predict\n",
      "    new_part = [predict_fn(shard) for shard in partition]\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in <listcomp>\n",
      "    new_part = [predict_fn(shard) for shard in partition]\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 448, in predict_fn\n",
      "    y = local_model.predict(shard[\"x\"], **params)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 130, in _method_wrapper\n",
      "    return method(self, *args, **kwargs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 1613, in predict\n",
      "    callbacks.on_predict_end()\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 582, in on_predict_end\n",
      "    callback.on_predict_end(logs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 979, in on_predict_end\n",
      "    self._finalize_progbar(logs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 1026, in _finalize_progbar\n",
      "    self.progbar.update(self.seen, list(logs.items()), finalize=True)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py\", line 581, in update\n",
      "    numdigits = int(np.log10(self.target)) + 1\n",
      "OverflowError: cannot convert float infinity to integer\n",
      "\n",
      "\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:452)\n",
      "\tat org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:588)\n",
      "\tat org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:571)\n",
      "\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:406)\n",
      "\tat org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)\n",
      "\tat org.apache.spark.storage.memory.MemoryStore.putIterator(MemoryStore.scala:221)\n",
      "\tat org.apache.spark.storage.memory.MemoryStore.putIteratorAsBytes(MemoryStore.scala:349)\n",
      "\tat org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:1182)\n",
      "\tat org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:1156)\n",
      "\tat org.apache.spark.storage.BlockManager.doPut(BlockManager.scala:1091)\n",
      "\tat org.apache.spark.storage.BlockManager.doPutIterator(BlockManager.scala:1156)\n",
      "\tat org.apache.spark.storage.BlockManager.getOrElseUpdate(BlockManager.scala:882)\n",
      "\tat org.apache.spark.rdd.RDD.getOrCompute(RDD.scala:335)\n",
      "\tat org.apache.spark.rdd.RDD.iterator(RDD.scala:286)\n",
      "\tat org.apache.spark.api.python.PythonRDD.compute(PythonRDD.scala:65)\n",
      "\tat org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)\n",
      "\tat org.apache.spark.rdd.RDD.iterator(RDD.scala:288)\n",
      "\tat org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)\n",
      "\tat org.apache.spark.scheduler.Task.run(Task.scala:121)\n",
      "\tat org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)\n",
      "\tat org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)\n",
      "\tat org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)\n",
      "\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n",
      "\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n",
      "\tat java.lang.Thread.run(Thread.java:748)\n",
      "2021-02-02 15:52:52 ERROR Executor:91 - Exception in task 1.0 in stage 10.0 (TID 41)\n",
      "org.apache.spark.api.python.PythonException: Traceback (most recent call last):\n",
      "  File \"/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/worker.py\", line 377, in main\n",
      "    process()\n",
      "  File \"/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/worker.py\", line 372, in process\n",
      "    serializer.dump_stream(func(split_index, iterator), outfile)\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/data/ray_rdd.py\", line 133, in <lambda>\n",
      "    lambda idx, _: get_from_ray(idx, address, password, meta_store_name))\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/data/ray_rdd.py\", line 99, in get_from_ray\n",
      "    partition = ray.get(object_id)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/ray/worker.py\", line 1513, in get\n",
      "    raise value.as_instanceof_cause()\n",
      "ray.exceptions.RayTaskError(OverflowError): \u001b[36mray::TFRunner.predict()\u001b[39m (pid=14333, ip=10.239.44.107)\n",
      "  File \"python/ray/_raylet.pyx\", line 452, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 407, in ray._raylet.execute_task.function_executor\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in predict\n",
      "    new_part = [predict_fn(shard) for shard in partition]\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in <listcomp>\n",
      "    new_part = [predict_fn(shard) for shard in partition]\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 448, in predict_fn\n",
      "    y = local_model.predict(shard[\"x\"], **params)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 130, in _method_wrapper\n",
      "    return method(self, *args, **kwargs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 1613, in predict\n",
      "    callbacks.on_predict_end()\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 582, in on_predict_end\n",
      "    callback.on_predict_end(logs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 979, in on_predict_end\n",
      "    self._finalize_progbar(logs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 1026, in _finalize_progbar\n",
      "    self.progbar.update(self.seen, list(logs.items()), finalize=True)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py\", line 581, in update\n",
      "    numdigits = int(np.log10(self.target)) + 1\n",
      "OverflowError: cannot convert float infinity to integer\n",
      "\n",
      "\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:452)\n",
      "\tat org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:588)\n",
      "\tat org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:571)\n",
      "\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:406)\n",
      "\tat org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)\n",
      "\tat org.apache.spark.storage.memory.MemoryStore.putIterator(MemoryStore.scala:221)\n",
      "\tat org.apache.spark.storage.memory.MemoryStore.putIteratorAsBytes(MemoryStore.scala:349)\n",
      "\tat org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:1182)\n",
      "\tat org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:1156)\n",
      "\tat org.apache.spark.storage.BlockManager.doPut(BlockManager.scala:1091)\n",
      "\tat org.apache.spark.storage.BlockManager.doPutIterator(BlockManager.scala:1156)\n",
      "\tat org.apache.spark.storage.BlockManager.getOrElseUpdate(BlockManager.scala:882)\n",
      "\tat org.apache.spark.rdd.RDD.getOrCompute(RDD.scala:335)\n",
      "\tat org.apache.spark.rdd.RDD.iterator(RDD.scala:286)\n",
      "\tat org.apache.spark.api.python.PythonRDD.compute(PythonRDD.scala:65)\n",
      "\tat org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)\n",
      "\tat org.apache.spark.rdd.RDD.iterator(RDD.scala:288)\n",
      "\tat org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)\n",
      "\tat org.apache.spark.scheduler.Task.run(Task.scala:121)\n",
      "\tat org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)\n",
      "\tat org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)\n",
      "\tat org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)\n",
      "\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n",
      "\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n",
      "\tat java.lang.Thread.run(Thread.java:748)\n",
      "2021-02-02 15:52:52 ERROR Executor:91 - Exception in task 1.0 in stage 10.0 (TID 41)\n",
      "org.apache.spark.api.python.PythonException: Traceback (most recent call last):\n",
      "  File \"/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/worker.py\", line 377, in main\n",
      "    process()\n",
      "  File \"/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/worker.py\", line 372, in process\n",
      "    serializer.dump_stream(func(split_index, iterator), outfile)\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/data/ray_rdd.py\", line 133, in <lambda>\n",
      "    lambda idx, _: get_from_ray(idx, address, password, meta_store_name))\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/data/ray_rdd.py\", line 99, in get_from_ray\n",
      "    partition = ray.get(object_id)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/ray/worker.py\", line 1513, in get\n",
      "    raise value.as_instanceof_cause()\n",
      "ray.exceptions.RayTaskError(OverflowError): \u001b[36mray::TFRunner.predict()\u001b[39m (pid=14333, ip=10.239.44.107)\n",
      "  File \"python/ray/_raylet.pyx\", line 452, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 407, in ray._raylet.execute_task.function_executor\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in predict\n",
      "    new_part = [predict_fn(shard) for shard in partition]\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in <listcomp>\n",
      "    new_part = [predict_fn(shard) for shard in partition]\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 448, in predict_fn\n",
      "    y = local_model.predict(shard[\"x\"], **params)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 130, in _method_wrapper\n",
      "    return method(self, *args, **kwargs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 1613, in predict\n",
      "    callbacks.on_predict_end()\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 582, in on_predict_end\n",
      "    callback.on_predict_end(logs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 979, in on_predict_end\n",
      "    self._finalize_progbar(logs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 1026, in _finalize_progbar\n",
      "    self.progbar.update(self.seen, list(logs.items()), finalize=True)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py\", line 581, in update\n",
      "    numdigits = int(np.log10(self.target)) + 1\n",
      "OverflowError: cannot convert float infinity to integer\n",
      "\n",
      "\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:452)\n",
      "\tat org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:588)\n",
      "\tat org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:571)\n",
      "\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:406)\n",
      "\tat org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)\n",
      "\tat org.apache.spark.storage.memory.MemoryStore.putIterator(MemoryStore.scala:221)\n",
      "\tat org.apache.spark.storage.memory.MemoryStore.putIteratorAsBytes(MemoryStore.scala:349)\n",
      "\tat org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:1182)\n",
      "\tat org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:1156)\n",
      "\tat org.apache.spark.storage.BlockManager.doPut(BlockManager.scala:1091)\n",
      "\tat org.apache.spark.storage.BlockManager.doPutIterator(BlockManager.scala:1156)\n",
      "\tat org.apache.spark.storage.BlockManager.getOrElseUpdate(BlockManager.scala:882)\n",
      "\tat org.apache.spark.rdd.RDD.getOrCompute(RDD.scala:335)\n",
      "\tat org.apache.spark.rdd.RDD.iterator(RDD.scala:286)\n",
      "\tat org.apache.spark.api.python.PythonRDD.compute(PythonRDD.scala:65)\n",
      "\tat org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)\n",
      "\tat org.apache.spark.rdd.RDD.iterator(RDD.scala:288)\n",
      "\tat org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)\n",
      "\tat org.apache.spark.scheduler.Task.run(Task.scala:121)\n",
      "\tat org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)\n",
      "\tat org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)\n",
      "\tat org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)\n",
      "\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n",
      "\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n",
      "\tat java.lang.Thread.run(Thread.java:748)\n",
      "2021-02-02 15:52:52 ERROR Executor:91 - Exception in task 1.0 in stage 10.0 (TID 41)\n",
      "org.apache.spark.api.python.PythonException: Traceback (most recent call last):\n",
      "  File \"/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/worker.py\", line 377, in main\n",
      "    process()\n",
      "  File \"/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/worker.py\", line 372, in process\n",
      "    serializer.dump_stream(func(split_index, iterator), outfile)\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/data/ray_rdd.py\", line 133, in <lambda>\n",
      "    lambda idx, _: get_from_ray(idx, address, password, meta_store_name))\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/data/ray_rdd.py\", line 99, in get_from_ray\n",
      "    partition = ray.get(object_id)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/ray/worker.py\", line 1513, in get\n",
      "    raise value.as_instanceof_cause()\n",
      "ray.exceptions.RayTaskError(OverflowError): \u001b[36mray::TFRunner.predict()\u001b[39m (pid=14333, ip=10.239.44.107)\n",
      "  File \"python/ray/_raylet.pyx\", line 452, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 407, in ray._raylet.execute_task.function_executor\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in predict\n",
      "    new_part = [predict_fn(shard) for shard in partition]\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in <listcomp>\n",
      "    new_part = [predict_fn(shard) for shard in partition]\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 448, in predict_fn\n",
      "    y = local_model.predict(shard[\"x\"], **params)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 130, in _method_wrapper\n",
      "    return method(self, *args, **kwargs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 1613, in predict\n",
      "    callbacks.on_predict_end()\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 582, in on_predict_end\n",
      "    callback.on_predict_end(logs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 979, in on_predict_end\n",
      "    self._finalize_progbar(logs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 1026, in _finalize_progbar\n",
      "    self.progbar.update(self.seen, list(logs.items()), finalize=True)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py\", line 581, in update\n",
      "    numdigits = int(np.log10(self.target)) + 1\n",
      "OverflowError: cannot convert float infinity to integer\n",
      "\n",
      "\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:452)\n",
      "\tat org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:588)\n",
      "\tat org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:571)\n",
      "\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:406)\n",
      "\tat org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)\n",
      "\tat org.apache.spark.storage.memory.MemoryStore.putIterator(MemoryStore.scala:221)\n",
      "\tat org.apache.spark.storage.memory.MemoryStore.putIteratorAsBytes(MemoryStore.scala:349)\n",
      "\tat org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:1182)\n",
      "\tat org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:1156)\n",
      "\tat org.apache.spark.storage.BlockManager.doPut(BlockManager.scala:1091)\n",
      "\tat org.apache.spark.storage.BlockManager.doPutIterator(BlockManager.scala:1156)\n",
      "\tat org.apache.spark.storage.BlockManager.getOrElseUpdate(BlockManager.scala:882)\n",
      "\tat org.apache.spark.rdd.RDD.getOrCompute(RDD.scala:335)\n",
      "\tat org.apache.spark.rdd.RDD.iterator(RDD.scala:286)\n",
      "\tat org.apache.spark.api.python.PythonRDD.compute(PythonRDD.scala:65)\n",
      "\tat org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)\n",
      "\tat org.apache.spark.rdd.RDD.iterator(RDD.scala:288)\n",
      "\tat org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)\n",
      "\tat org.apache.spark.scheduler.Task.run(Task.scala:121)\n",
      "\tat org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)\n",
      "\tat org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)\n",
      "\tat org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)\n",
      "\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n",
      "\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n",
      "\tat java.lang.Thread.run(Thread.java:748)\n",
      "2021-02-02 15:52:52 ERROR TaskSetManager:70 - Task 1 in stage 10.0 failed 1 times; aborting job\n",
      "2021-02-02 15:52:52 ERROR TaskSetManager:70 - Task 1 in stage 10.0 failed 1 times; aborting job\n",
      "2021-02-02 15:52:52 ERROR TaskSetManager:70 - Task 1 in stage 10.0 failed 1 times; aborting job\n",
      "2021-02-02 15:52:52 ERROR TaskSetManager:70 - Task 1 in stage 10.0 failed 1 times; aborting job\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(pid=14333)\u001b[0m /home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py:581: RuntimeWarning: divide by zero encountered in log10\n",
      "\u001b[2m\u001b[36m(pid=14333)\u001b[0m   numdigits = int(np.log10(self.target)) + 1\n",
      "\u001b[2m\u001b[36m(pid=14333)\u001b[0m \r\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(pid=14333)\u001b[0m /home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py:581: RuntimeWarning: divide by zero encountered in log10\n",
      "\u001b[2m\u001b[36m(pid=14333)\u001b[0m   numdigits = int(np.log10(self.target)) + 1\n",
      "\u001b[2m\u001b[36m(pid=14333)\u001b[0m \r\n",
      "\u001b[2m\u001b[36m(pid=14333)\u001b[0m /home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py:581: RuntimeWarning: divide by zero encountered in log10\n",
      "\u001b[2m\u001b[36m(pid=14333)\u001b[0m   numdigits = int(np.log10(self.target)) + 1\n",
      "\u001b[2m\u001b[36m(pid=14333)\u001b[0m \r\n"
     ]
    },
    {
     "ename": "Py4JJavaError",
     "evalue": "An error occurred while calling z:org.apache.spark.api.python.PythonRDD.collectAndServe.\n: org.apache.spark.SparkException: Job aborted due to stage failure: Task 1 in stage 10.0 failed 1 times, most recent failure: Lost task 1.0 in stage 10.0 (TID 41, localhost, executor driver): org.apache.spark.api.python.PythonException: Traceback (most recent call last):\n  File \"/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/worker.py\", line 377, in main\n    process()\n  File \"/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/worker.py\", line 372, in process\n    serializer.dump_stream(func(split_index, iterator), outfile)\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/data/ray_rdd.py\", line 133, in <lambda>\n    lambda idx, _: get_from_ray(idx, address, password, meta_store_name))\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/data/ray_rdd.py\", line 99, in get_from_ray\n    partition = ray.get(object_id)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/ray/worker.py\", line 1513, in get\n    raise value.as_instanceof_cause()\nray.exceptions.RayTaskError(OverflowError): \u001b[36mray::TFRunner.predict()\u001b[39m (pid=14333, ip=10.239.44.107)\n  File \"python/ray/_raylet.pyx\", line 452, in ray._raylet.execute_task\n  File \"python/ray/_raylet.pyx\", line 407, in ray._raylet.execute_task.function_executor\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in predict\n    new_part = [predict_fn(shard) for shard in partition]\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in <listcomp>\n    new_part = [predict_fn(shard) for shard in partition]\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 448, in predict_fn\n    y = local_model.predict(shard[\"x\"], **params)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 130, in _method_wrapper\n    return method(self, *args, **kwargs)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 1613, in predict\n    callbacks.on_predict_end()\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 582, in on_predict_end\n    callback.on_predict_end(logs)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 979, in on_predict_end\n    self._finalize_progbar(logs)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 1026, in _finalize_progbar\n    self.progbar.update(self.seen, list(logs.items()), finalize=True)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py\", line 581, in update\n    numdigits = int(np.log10(self.target)) + 1\nOverflowError: cannot convert float infinity to integer\n\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:452)\n\tat org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:588)\n\tat org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:571)\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:406)\n\tat org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)\n\tat org.apache.spark.storage.memory.MemoryStore.putIterator(MemoryStore.scala:221)\n\tat org.apache.spark.storage.memory.MemoryStore.putIteratorAsBytes(MemoryStore.scala:349)\n\tat org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:1182)\n\tat org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:1156)\n\tat org.apache.spark.storage.BlockManager.doPut(BlockManager.scala:1091)\n\tat org.apache.spark.storage.BlockManager.doPutIterator(BlockManager.scala:1156)\n\tat org.apache.spark.storage.BlockManager.getOrElseUpdate(BlockManager.scala:882)\n\tat org.apache.spark.rdd.RDD.getOrCompute(RDD.scala:335)\n\tat org.apache.spark.rdd.RDD.iterator(RDD.scala:286)\n\tat org.apache.spark.api.python.PythonRDD.compute(PythonRDD.scala:65)\n\tat org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)\n\tat org.apache.spark.rdd.RDD.iterator(RDD.scala:288)\n\tat org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)\n\tat org.apache.spark.scheduler.Task.run(Task.scala:121)\n\tat org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)\n\tat org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)\n\tat org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n\tat java.lang.Thread.run(Thread.java:748)\n\nDriver stacktrace:\n\tat org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1889)\n\tat org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1877)\n\tat org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1876)\n\tat scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)\n\tat scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)\n\tat org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1876)\n\tat org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:926)\n\tat org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:926)\n\tat scala.Option.foreach(Option.scala:257)\n\tat org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:926)\n\tat org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2110)\n\tat org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2059)\n\tat org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2048)\n\tat org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)\n\tat org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:737)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2061)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2082)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2101)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2126)\n\tat org.apache.spark.rdd.RDD$$anonfun$collect$1.apply(RDD.scala:945)\n\tat org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)\n\tat org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)\n\tat org.apache.spark.rdd.RDD.withScope(RDD.scala:363)\n\tat org.apache.spark.rdd.RDD.collect(RDD.scala:944)\n\tat org.apache.spark.api.python.PythonRDD$.collectAndServe(PythonRDD.scala:166)\n\tat org.apache.spark.api.python.PythonRDD.collectAndServe(PythonRDD.scala)\n\tat sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n\tat sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\n\tat sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n\tat java.lang.reflect.Method.invoke(Method.java:498)\n\tat py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)\n\tat py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)\n\tat py4j.Gateway.invoke(Gateway.java:282)\n\tat py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)\n\tat py4j.commands.CallCommand.execute(CallCommand.java:79)\n\tat py4j.GatewayConnection.run(GatewayConnection.java:238)\n\tat java.lang.Thread.run(Thread.java:748)\nCaused by: org.apache.spark.api.python.PythonException: Traceback (most recent call last):\n  File \"/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/worker.py\", line 377, in main\n    process()\n  File \"/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/worker.py\", line 372, in process\n    serializer.dump_stream(func(split_index, iterator), outfile)\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/data/ray_rdd.py\", line 133, in <lambda>\n    lambda idx, _: get_from_ray(idx, address, password, meta_store_name))\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/data/ray_rdd.py\", line 99, in get_from_ray\n    partition = ray.get(object_id)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/ray/worker.py\", line 1513, in get\n    raise value.as_instanceof_cause()\nray.exceptions.RayTaskError(OverflowError): \u001b[36mray::TFRunner.predict()\u001b[39m (pid=14333, ip=10.239.44.107)\n  File \"python/ray/_raylet.pyx\", line 452, in ray._raylet.execute_task\n  File \"python/ray/_raylet.pyx\", line 407, in ray._raylet.execute_task.function_executor\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in predict\n    new_part = [predict_fn(shard) for shard in partition]\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in <listcomp>\n    new_part = [predict_fn(shard) for shard in partition]\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 448, in predict_fn\n    y = local_model.predict(shard[\"x\"], **params)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 130, in _method_wrapper\n    return method(self, *args, **kwargs)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 1613, in predict\n    callbacks.on_predict_end()\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 582, in on_predict_end\n    callback.on_predict_end(logs)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 979, in on_predict_end\n    self._finalize_progbar(logs)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 1026, in _finalize_progbar\n    self.progbar.update(self.seen, list(logs.items()), finalize=True)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py\", line 581, in update\n    numdigits = int(np.log10(self.target)) + 1\nOverflowError: cannot convert float infinity to integer\n\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:452)\n\tat org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:588)\n\tat org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:571)\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:406)\n\tat org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)\n\tat org.apache.spark.storage.memory.MemoryStore.putIterator(MemoryStore.scala:221)\n\tat org.apache.spark.storage.memory.MemoryStore.putIteratorAsBytes(MemoryStore.scala:349)\n\tat org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:1182)\n\tat org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:1156)\n\tat org.apache.spark.storage.BlockManager.doPut(BlockManager.scala:1091)\n\tat org.apache.spark.storage.BlockManager.doPutIterator(BlockManager.scala:1156)\n\tat org.apache.spark.storage.BlockManager.getOrElseUpdate(BlockManager.scala:882)\n\tat org.apache.spark.rdd.RDD.getOrCompute(RDD.scala:335)\n\tat org.apache.spark.rdd.RDD.iterator(RDD.scala:286)\n\tat org.apache.spark.api.python.PythonRDD.compute(PythonRDD.scala:65)\n\tat org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)\n\tat org.apache.spark.rdd.RDD.iterator(RDD.scala:288)\n\tat org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)\n\tat org.apache.spark.scheduler.Task.run(Task.scala:121)\n\tat org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)\n\tat org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)\n\tat org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n\t... 1 more\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mPy4JJavaError\u001b[0m                             Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-19-c02268e07d6c>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     31\u001b[0m             \u001b[0;31m#sample_shards = sample_shards.transform_shard(transform_to_dict)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     32\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 33\u001b[0;31m             \u001b[0mpreds\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mest\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpredict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msample_shards\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     34\u001b[0m             \u001b[0mnext_index\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msample\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpreds\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtemperature\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     35\u001b[0m             \u001b[0mnext_char\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mchars\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mnext_index\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/analytics-zoo/pyzoo/zoo/orca/learn/tf2/estimator.py\u001b[0m in \u001b[0;36mpredict\u001b[0;34m(self, data, batch_size, verbose, steps, callbacks, data_config, feature_cols)\u001b[0m\n\u001b[1;32m    303\u001b[0m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mconvert_predict_xshards_to_dataframe\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpred_shards\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    304\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mSparkXShards\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 305\u001b[0;31m             \u001b[0mpred_shards\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_predict_spark_xshards\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mparams\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    306\u001b[0m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mupdate_predict_xshards\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpred_shards\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    307\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/analytics-zoo/pyzoo/zoo/orca/learn/tf2/estimator.py\u001b[0m in \u001b[0;36m_predict_spark_xshards\u001b[0;34m(self, xshards, params)\u001b[0m\n\u001b[1;32m    276\u001b[0m                                                                \u001b[0mtransform_func\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    277\u001b[0m                                                                gang_scheduling=False)\n\u001b[0;32m--> 278\u001b[0;31m         \u001b[0mspark_xshards\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpred_shards\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_spark_xshards\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    279\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mspark_xshards\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    280\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/analytics-zoo/pyzoo/zoo/orca/data/shard.py\u001b[0m in \u001b[0;36mto_spark_xshards\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    160\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    161\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mto_spark_xshards\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 162\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mSparkXShards\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mray_rdd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_spark_rdd\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    163\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    164\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mstaticmethod\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/analytics-zoo/pyzoo/zoo/orca/data/shard.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, rdd, transient)\u001b[0m\n\u001b[1;32m    182\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrdd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcache\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    183\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0meager\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 184\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    185\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtype\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    186\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/analytics-zoo/pyzoo/zoo/orca/data/shard.py\u001b[0m in \u001b[0;36mcompute\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    238\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    239\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mcompute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 240\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrdd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcount\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    241\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    242\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/rdd.py\u001b[0m in \u001b[0;36mcount\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1053\u001b[0m         \u001b[0;36m3\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1054\u001b[0m         \"\"\"\n\u001b[0;32m-> 1055\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmapPartitions\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0m_\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1056\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1057\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mstats\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/rdd.py\u001b[0m in \u001b[0;36msum\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1044\u001b[0m         \u001b[0;36m6.0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1045\u001b[0m         \"\"\"\n\u001b[0;32m-> 1046\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmapPartitions\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfold\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moperator\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madd\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1047\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1048\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mcount\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/rdd.py\u001b[0m in \u001b[0;36mfold\u001b[0;34m(self, zeroValue, op)\u001b[0m\n\u001b[1;32m    915\u001b[0m         \u001b[0;31m# zeroValue provided to each partition is unique from the one provided\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    916\u001b[0m         \u001b[0;31m# to the final reduce call\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 917\u001b[0;31m         \u001b[0mvals\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmapPartitions\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfunc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcollect\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    918\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mreduce\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mop\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvals\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mzeroValue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    919\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/rdd.py\u001b[0m in \u001b[0;36mcollect\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    814\u001b[0m         \"\"\"\n\u001b[1;32m    815\u001b[0m         \u001b[0;32mwith\u001b[0m \u001b[0mSCCallSiteSync\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcontext\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mcss\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 816\u001b[0;31m             \u001b[0msock_info\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mctx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jvm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mPythonRDD\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcollectAndServe\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jrdd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrdd\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    817\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_load_from_socket\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msock_info\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jrdd_deserializer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    818\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/py4j-0.10.7-src.zip/py4j/java_gateway.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args)\u001b[0m\n\u001b[1;32m   1255\u001b[0m         \u001b[0manswer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgateway_client\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend_command\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcommand\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1256\u001b[0m         return_value = get_return_value(\n\u001b[0;32m-> 1257\u001b[0;31m             answer, self.gateway_client, self.target_id, self.name)\n\u001b[0m\u001b[1;32m   1258\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1259\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mtemp_arg\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mtemp_args\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/py4j-0.10.7-src.zip/py4j/protocol.py\u001b[0m in \u001b[0;36mget_return_value\u001b[0;34m(answer, gateway_client, target_id, name)\u001b[0m\n\u001b[1;32m    326\u001b[0m                 raise Py4JJavaError(\n\u001b[1;32m    327\u001b[0m                     \u001b[0;34m\"An error occurred while calling {0}{1}{2}.\\n\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 328\u001b[0;31m                     format(target_id, \".\", name), value)\n\u001b[0m\u001b[1;32m    329\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    330\u001b[0m                 raise Py4JError(\n",
      "\u001b[0;31mPy4JJavaError\u001b[0m: An error occurred while calling z:org.apache.spark.api.python.PythonRDD.collectAndServe.\n: org.apache.spark.SparkException: Job aborted due to stage failure: Task 1 in stage 10.0 failed 1 times, most recent failure: Lost task 1.0 in stage 10.0 (TID 41, localhost, executor driver): org.apache.spark.api.python.PythonException: Traceback (most recent call last):\n  File \"/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/worker.py\", line 377, in main\n    process()\n  File \"/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/worker.py\", line 372, in process\n    serializer.dump_stream(func(split_index, iterator), outfile)\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/data/ray_rdd.py\", line 133, in <lambda>\n    lambda idx, _: get_from_ray(idx, address, password, meta_store_name))\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/data/ray_rdd.py\", line 99, in get_from_ray\n    partition = ray.get(object_id)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/ray/worker.py\", line 1513, in get\n    raise value.as_instanceof_cause()\nray.exceptions.RayTaskError(OverflowError): \u001b[36mray::TFRunner.predict()\u001b[39m (pid=14333, ip=10.239.44.107)\n  File \"python/ray/_raylet.pyx\", line 452, in ray._raylet.execute_task\n  File \"python/ray/_raylet.pyx\", line 407, in ray._raylet.execute_task.function_executor\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in predict\n    new_part = [predict_fn(shard) for shard in partition]\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in <listcomp>\n    new_part = [predict_fn(shard) for shard in partition]\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 448, in predict_fn\n    y = local_model.predict(shard[\"x\"], **params)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 130, in _method_wrapper\n    return method(self, *args, **kwargs)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 1613, in predict\n    callbacks.on_predict_end()\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 582, in on_predict_end\n    callback.on_predict_end(logs)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 979, in on_predict_end\n    self._finalize_progbar(logs)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 1026, in _finalize_progbar\n    self.progbar.update(self.seen, list(logs.items()), finalize=True)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py\", line 581, in update\n    numdigits = int(np.log10(self.target)) + 1\nOverflowError: cannot convert float infinity to integer\n\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:452)\n\tat org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:588)\n\tat org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:571)\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:406)\n\tat org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)\n\tat org.apache.spark.storage.memory.MemoryStore.putIterator(MemoryStore.scala:221)\n\tat org.apache.spark.storage.memory.MemoryStore.putIteratorAsBytes(MemoryStore.scala:349)\n\tat org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:1182)\n\tat org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:1156)\n\tat org.apache.spark.storage.BlockManager.doPut(BlockManager.scala:1091)\n\tat org.apache.spark.storage.BlockManager.doPutIterator(BlockManager.scala:1156)\n\tat org.apache.spark.storage.BlockManager.getOrElseUpdate(BlockManager.scala:882)\n\tat org.apache.spark.rdd.RDD.getOrCompute(RDD.scala:335)\n\tat org.apache.spark.rdd.RDD.iterator(RDD.scala:286)\n\tat org.apache.spark.api.python.PythonRDD.compute(PythonRDD.scala:65)\n\tat org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)\n\tat org.apache.spark.rdd.RDD.iterator(RDD.scala:288)\n\tat org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)\n\tat org.apache.spark.scheduler.Task.run(Task.scala:121)\n\tat org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)\n\tat org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)\n\tat org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n\tat java.lang.Thread.run(Thread.java:748)\n\nDriver stacktrace:\n\tat org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1889)\n\tat org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1877)\n\tat org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1876)\n\tat scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)\n\tat scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)\n\tat org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1876)\n\tat org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:926)\n\tat org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:926)\n\tat scala.Option.foreach(Option.scala:257)\n\tat org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:926)\n\tat org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2110)\n\tat org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2059)\n\tat org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2048)\n\tat org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)\n\tat org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:737)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2061)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2082)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2101)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2126)\n\tat org.apache.spark.rdd.RDD$$anonfun$collect$1.apply(RDD.scala:945)\n\tat org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)\n\tat org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)\n\tat org.apache.spark.rdd.RDD.withScope(RDD.scala:363)\n\tat org.apache.spark.rdd.RDD.collect(RDD.scala:944)\n\tat org.apache.spark.api.python.PythonRDD$.collectAndServe(PythonRDD.scala:166)\n\tat org.apache.spark.api.python.PythonRDD.collectAndServe(PythonRDD.scala)\n\tat sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n\tat sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\n\tat sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n\tat java.lang.reflect.Method.invoke(Method.java:498)\n\tat py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)\n\tat py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)\n\tat py4j.Gateway.invoke(Gateway.java:282)\n\tat py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)\n\tat py4j.commands.CallCommand.execute(CallCommand.java:79)\n\tat py4j.GatewayConnection.run(GatewayConnection.java:238)\n\tat java.lang.Thread.run(Thread.java:748)\nCaused by: org.apache.spark.api.python.PythonException: Traceback (most recent call last):\n  File \"/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/worker.py\", line 377, in main\n    process()\n  File \"/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/worker.py\", line 372, in process\n    serializer.dump_stream(func(split_index, iterator), outfile)\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/data/ray_rdd.py\", line 133, in <lambda>\n    lambda idx, _: get_from_ray(idx, address, password, meta_store_name))\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/data/ray_rdd.py\", line 99, in get_from_ray\n    partition = ray.get(object_id)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/ray/worker.py\", line 1513, in get\n    raise value.as_instanceof_cause()\nray.exceptions.RayTaskError(OverflowError): \u001b[36mray::TFRunner.predict()\u001b[39m (pid=14333, ip=10.239.44.107)\n  File \"python/ray/_raylet.pyx\", line 452, in ray._raylet.execute_task\n  File \"python/ray/_raylet.pyx\", line 407, in ray._raylet.execute_task.function_executor\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in predict\n    new_part = [predict_fn(shard) for shard in partition]\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in <listcomp>\n    new_part = [predict_fn(shard) for shard in partition]\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 448, in predict_fn\n    y = local_model.predict(shard[\"x\"], **params)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 130, in _method_wrapper\n    return method(self, *args, **kwargs)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 1613, in predict\n    callbacks.on_predict_end()\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 582, in on_predict_end\n    callback.on_predict_end(logs)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 979, in on_predict_end\n    self._finalize_progbar(logs)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 1026, in _finalize_progbar\n    self.progbar.update(self.seen, list(logs.items()), finalize=True)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py\", line 581, in update\n    numdigits = int(np.log10(self.target)) + 1\nOverflowError: cannot convert float infinity to integer\n\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:452)\n\tat org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:588)\n\tat org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:571)\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:406)\n\tat org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)\n\tat org.apache.spark.storage.memory.MemoryStore.putIterator(MemoryStore.scala:221)\n\tat org.apache.spark.storage.memory.MemoryStore.putIteratorAsBytes(MemoryStore.scala:349)\n\tat org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:1182)\n\tat org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:1156)\n\tat org.apache.spark.storage.BlockManager.doPut(BlockManager.scala:1091)\n\tat org.apache.spark.storage.BlockManager.doPutIterator(BlockManager.scala:1156)\n\tat org.apache.spark.storage.BlockManager.getOrElseUpdate(BlockManager.scala:882)\n\tat org.apache.spark.rdd.RDD.getOrCompute(RDD.scala:335)\n\tat org.apache.spark.rdd.RDD.iterator(RDD.scala:286)\n\tat org.apache.spark.api.python.PythonRDD.compute(PythonRDD.scala:65)\n\tat org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)\n\tat org.apache.spark.rdd.RDD.iterator(RDD.scala:288)\n\tat org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)\n\tat org.apache.spark.scheduler.Task.run(Task.scala:121)\n\tat org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)\n\tat org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)\n\tat org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n\t... 1 more\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 10:==============>                                           (1 + 2) / 4]\u001b[2m\u001b[36m(pid=14333)\u001b[0m \n",
      "[Stage 10:==============>                                           (1 + 1) / 4]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(pid=14333)\u001b[0m \n",
      "\u001b[2m\u001b[36m(pid=14333)\u001b[0m \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-02-02 15:52:57,190\tERROR worker.py:1012 -- Possible unhandled error from worker: \u001b[36mray::TFRunner.predict()\u001b[39m (pid=14333, ip=10.239.44.107)\n",
      "  File \"python/ray/_raylet.pyx\", line 452, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 407, in ray._raylet.execute_task.function_executor\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in predict\n",
      "    new_part = [predict_fn(shard) for shard in partition]\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in <listcomp>\n",
      "    new_part = [predict_fn(shard) for shard in partition]\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 448, in predict_fn\n",
      "    y = local_model.predict(shard[\"x\"], **params)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 130, in _method_wrapper\n",
      "    return method(self, *args, **kwargs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 1613, in predict\n",
      "    callbacks.on_predict_end()\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 582, in on_predict_end\n",
      "    callback.on_predict_end(logs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 979, in on_predict_end\n",
      "    self._finalize_progbar(logs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 1026, in _finalize_progbar\n",
      "    self.progbar.update(self.seen, list(logs.items()), finalize=True)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py\", line 581, in update\n",
      "    numdigits = int(np.log10(self.target)) + 1\n",
      "OverflowError: cannot convert float infinity to integer\n",
      "2021-02-02 15:52:58,192\tERROR worker.py:1012 -- Possible unhandled error from worker: \u001b[36mray::TFRunner.predict()\u001b[39m (pid=14333, ip=10.239.44.107)\n",
      "  File \"python/ray/_raylet.pyx\", line 452, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 407, in ray._raylet.execute_task.function_executor\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in predict\n",
      "    new_part = [predict_fn(shard) for shard in partition]\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in <listcomp>\n",
      "    new_part = [predict_fn(shard) for shard in partition]\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 448, in predict_fn\n",
      "    y = local_model.predict(shard[\"x\"], **params)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 130, in _method_wrapper\n",
      "    return method(self, *args, **kwargs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 1613, in predict\n",
      "    callbacks.on_predict_end()\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 582, in on_predict_end\n",
      "    callback.on_predict_end(logs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 979, in on_predict_end\n",
      "    self._finalize_progbar(logs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 1026, in _finalize_progbar\n",
      "    self.progbar.update(self.seen, list(logs.items()), finalize=True)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py\", line 581, in update\n",
      "    numdigits = int(np.log10(self.target)) + 1\n",
      "OverflowError: cannot convert float infinity to integer\n",
      "2021-02-02 15:52:58,194\tERROR worker.py:1012 -- Possible unhandled error from worker: \u001b[36mray::TFRunner.predict()\u001b[39m (pid=14333, ip=10.239.44.107)\n",
      "  File \"python/ray/_raylet.pyx\", line 452, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 407, in ray._raylet.execute_task.function_executor\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in predict\n",
      "    new_part = [predict_fn(shard) for shard in partition]\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in <listcomp>\n",
      "    new_part = [predict_fn(shard) for shard in partition]\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 448, in predict_fn\n",
      "    y = local_model.predict(shard[\"x\"], **params)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 130, in _method_wrapper\n",
      "    return method(self, *args, **kwargs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 1613, in predict\n",
      "    callbacks.on_predict_end()\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 582, in on_predict_end\n",
      "    callback.on_predict_end(logs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 979, in on_predict_end\n",
      "    self._finalize_progbar(logs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 1026, in _finalize_progbar\n",
      "    self.progbar.update(self.seen, list(logs.items()), finalize=True)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py\", line 581, in update\n",
      "    numdigits = int(np.log10(self.target)) + 1\n",
      "OverflowError: cannot convert float infinity to integer\n"
     ]
    }
   ],
   "source": [
    "from zoo.orca.data import XShards\n",
    "\n",
    "def transform_to_dict(data):\n",
    "    return {\"x\": data[0]}\n",
    "\n",
    "\n",
    "for epoch in range(1, 60):\n",
    "    print('epoch', epoch)\n",
    "    # Fit the model for 1 epoch on the available training data\n",
    "    stats = est.fit(train_data_creator, \n",
    "                    epochs=1,\n",
    "                    batch_size=batch_size,\n",
    "                    steps_per_epoch=200278 // batch_size,\n",
    "                    verbose=0)\n",
    "    # Select a text seed at random\n",
    "    start_index = random.randint(0, len(text) - maxlen - 1)\n",
    "    generated_text = text[start_index: start_index + maxlen]\n",
    "    print('--- Generating with seed: \"' + generated_text + '\"')\n",
    "\n",
    "    for temperature in [0.2, 0.5, 1.0, 1.2]:\n",
    "        print('------ temperature:', temperature)\n",
    "        sys.stdout.write(generated_text)\n",
    "\n",
    "        # We generate 400 characters\n",
    "        for i in range(400):\n",
    "            sampled = np.zeros((1, maxlen, len(chars)))\n",
    "            for t, char in enumerate(generated_text):\n",
    "                sampled[0, t, char_indices[char]] = 1.\n",
    "                \n",
    "            sample_shards = XShards.partition({\"x\": sampled})\n",
    "            #sample_shards = sample_shards.transform_shard(transform_to_dict)\n",
    "\n",
    "            preds = est.predict(sample_shards)\n",
    "            next_index = sample(preds, temperature)\n",
    "            next_char = chars[next_index]\n",
    "\n",
    "            generated_text += next_char\n",
    "            generated_text = generated_text[1:]\n",
    "\n",
    "            sys.stdout.write(next_char)\n",
    "            sys.stdout.flush()\n",
    "        print()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "muslim-correspondence",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(pid=14333)\u001b[0m 2021-02-02 16:13:41.082677: W tensorflow/core/common_runtime/eager/context.cc:566] Unable to destroy server_ object, so releasing instead. Servers don't support clean shutdown.\n"
     ]
    }
   ],
   "source": [
    "est.shutdown()\n",
    "#stop_orca_context()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "incorporate-modeling",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- Generating with seed: \"y regard him on\n",
      "that account with some apprehension, they wo\"\n",
      "------ temperature: 0.2\n",
      "y regard him on\n",
      "that account with some apprehension, they wo------ temperature: 0.5\n",
      "y regard him on\n",
      "that account with some apprehension, they wo------ temperature: 1.0\n",
      "y regard him on\n",
      "that account with some apprehension, they wo------ temperature: 1.2\n",
      "y regard him on\n",
      "that account with some apprehension, they wo"
     ]
    }
   ],
   "source": [
    "from zoo.orca.data import XShards\n",
    "\n",
    "start_index = random.randint(0, len(text) - maxlen - 1)\n",
    "generated_text = text[start_index: start_index + maxlen]\n",
    "print('--- Generating with seed: \"' + generated_text + '\"')\n",
    "\n",
    "for temperature in [0.2, 0.5, 1.0, 1.2]:\n",
    "    print('------ temperature:', temperature)\n",
    "    sys.stdout.write(generated_text)\n",
    "\n",
    "    # We generate 400 characters\n",
    "    for i in range(400):\n",
    "        sampled = np.zeros((1, maxlen, len(chars)))\n",
    "        for t, char in enumerate(generated_text):\n",
    "            sampled[0, t, char_indices[char]] = 1.\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "beautiful-teddy",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\r",
      "[Stage 0:>                                                          (0 + 4) / 4]\r",
      "                                                                                \r"
     ]
    }
   ],
   "source": [
    "sample_shards = XShards.partition({\"x\": sampled})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "alpine-honey",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-02-02 20:27:32,645\tWARNING worker.py:792 -- When connecting to an existing cluster, _internal_config must match the cluster's _internal_config.\n",
      "2021-02-02 20:27:32,697\tWARNING worker.py:792 -- When connecting to an existing cluster, _internal_config must match the cluster's _internal_config.\n",
      "2021-02-02 20:27:32,795\tWARNING worker.py:792 -- When connecting to an existing cluster, _internal_config must match the cluster's _internal_config.\n",
      "2021-02-02 20:27:32,807\tWARNING worker.py:792 -- When connecting to an existing cluster, _internal_config must match the cluster's _internal_config.\n",
      "2021-02-02 20:27:32,907\tWARNING worker.py:792 -- When connecting to an existing cluster, _internal_config must match the cluster's _internal_config.\n",
      "2021-02-02 20:27:32,907\tWARNING worker.py:792 -- When connecting to an existing cluster, _internal_config must match the cluster's _internal_config.\n",
      "2021-02-02 20:27:32,907\tWARNING worker.py:792 -- When connecting to an existing cluster, _internal_config must match the cluster's _internal_config.\n",
      "2021-02-02 20:27:32,924\tWARNING worker.py:792 -- When connecting to an existing cluster, _internal_config must match the cluster's _internal_config.\n",
      "1/1 [==============================] - 0s 188us/step\n",
      "1/1 [==============================] - 0s 188us/step\n",
      "1/1 [==============================] - 0s 188us/step\n",
      "[Stage 2:==============>                                            (1 + 3) / 4]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(pid=18170)\u001b[0m \r",
      "1/1 [==============================] - ETA: 0s\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\r",
      "1/1 [==============================] - 0s 188us/step\n",
      "2021-02-02 20:27:33 ERROR Executor:91 - Exception in task 1.0 in stage 2.0 (TID 9)\n",
      "org.apache.spark.api.python.PythonException: Traceback (most recent call last):\n",
      "  File \"/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/worker.py\", line 377, in main\n",
      "    process()\n",
      "  File \"/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/worker.py\", line 372, in process\n",
      "    serializer.dump_stream(func(split_index, iterator), outfile)\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/data/ray_rdd.py\", line 133, in <lambda>\n",
      "    lambda idx, _: get_from_ray(idx, address, password, meta_store_name))\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/data/ray_rdd.py\", line 99, in get_from_ray\n",
      "    partition = ray.get(object_id)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/ray/worker.py\", line 1513, in get\n",
      "    raise value.as_instanceof_cause()\n",
      "ray.exceptions.RayTaskError(OverflowError): \u001b[36mray::TFRunner.predict()\u001b[39m (pid=18170, ip=10.239.44.107)\n",
      "  File \"python/ray/_raylet.pyx\", line 452, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 407, in ray._raylet.execute_task.function_executor\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in predict\n",
      "    new_part = [predict_fn(shard) for shard in partition]\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in <listcomp>\n",
      "    new_part = [predict_fn(shard) for shard in partition]\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 448, in predict_fn\n",
      "    y = local_model.predict(shard[\"x\"], **params)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 130, in _method_wrapper\n",
      "    return method(self, *args, **kwargs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 1613, in predict\n",
      "    callbacks.on_predict_end()\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 582, in on_predict_end\n",
      "    callback.on_predict_end(logs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 979, in on_predict_end\n",
      "    self._finalize_progbar(logs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 1026, in _finalize_progbar\n",
      "    self.progbar.update(self.seen, list(logs.items()), finalize=True)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py\", line 581, in update\n",
      "    numdigits = int(np.log10(self.target)) + 1\n",
      "OverflowError: cannot convert float infinity to integer\n",
      "\n",
      "\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:452)\n",
      "\tat org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:588)\n",
      "\tat org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:571)\n",
      "\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:406)\n",
      "\tat org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)\n",
      "\tat org.apache.spark.storage.memory.MemoryStore.putIterator(MemoryStore.scala:221)\n",
      "\tat org.apache.spark.storage.memory.MemoryStore.putIteratorAsBytes(MemoryStore.scala:349)\n",
      "\tat org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:1182)\n",
      "\tat org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:1156)\n",
      "\tat org.apache.spark.storage.BlockManager.doPut(BlockManager.scala:1091)\n",
      "\tat org.apache.spark.storage.BlockManager.doPutIterator(BlockManager.scala:1156)\n",
      "\tat org.apache.spark.storage.BlockManager.getOrElseUpdate(BlockManager.scala:882)\n",
      "\tat org.apache.spark.rdd.RDD.getOrCompute(RDD.scala:335)\n",
      "\tat org.apache.spark.rdd.RDD.iterator(RDD.scala:286)\n",
      "\tat org.apache.spark.api.python.PythonRDD.compute(PythonRDD.scala:65)\n",
      "\tat org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)\n",
      "\tat org.apache.spark.rdd.RDD.iterator(RDD.scala:288)\n",
      "\tat org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)\n",
      "\tat org.apache.spark.scheduler.Task.run(Task.scala:121)\n",
      "\tat org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)\n",
      "\tat org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)\n",
      "\tat org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)\n",
      "\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n",
      "\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n",
      "\tat java.lang.Thread.run(Thread.java:748)\n",
      "2021-02-02 20:27:33 ERROR Executor:91 - Exception in task 1.0 in stage 2.0 (TID 9)\n",
      "org.apache.spark.api.python.PythonException: Traceback (most recent call last):\n",
      "  File \"/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/worker.py\", line 377, in main\n",
      "    process()\n",
      "  File \"/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/worker.py\", line 372, in process\n",
      "    serializer.dump_stream(func(split_index, iterator), outfile)\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/data/ray_rdd.py\", line 133, in <lambda>\n",
      "    lambda idx, _: get_from_ray(idx, address, password, meta_store_name))\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/data/ray_rdd.py\", line 99, in get_from_ray\n",
      "    partition = ray.get(object_id)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/ray/worker.py\", line 1513, in get\n",
      "    raise value.as_instanceof_cause()\n",
      "ray.exceptions.RayTaskError(OverflowError): \u001b[36mray::TFRunner.predict()\u001b[39m (pid=18170, ip=10.239.44.107)\n",
      "  File \"python/ray/_raylet.pyx\", line 452, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 407, in ray._raylet.execute_task.function_executor\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in predict\n",
      "    new_part = [predict_fn(shard) for shard in partition]\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in <listcomp>\n",
      "    new_part = [predict_fn(shard) for shard in partition]\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 448, in predict_fn\n",
      "    y = local_model.predict(shard[\"x\"], **params)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 130, in _method_wrapper\n",
      "    return method(self, *args, **kwargs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 1613, in predict\n",
      "    callbacks.on_predict_end()\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 582, in on_predict_end\n",
      "    callback.on_predict_end(logs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 979, in on_predict_end\n",
      "    self._finalize_progbar(logs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 1026, in _finalize_progbar\n",
      "    self.progbar.update(self.seen, list(logs.items()), finalize=True)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py\", line 581, in update\n",
      "    numdigits = int(np.log10(self.target)) + 1\n",
      "OverflowError: cannot convert float infinity to integer\n",
      "\n",
      "\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:452)\n",
      "\tat org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:588)\n",
      "\tat org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:571)\n",
      "\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:406)\n",
      "\tat org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)\n",
      "\tat org.apache.spark.storage.memory.MemoryStore.putIterator(MemoryStore.scala:221)\n",
      "\tat org.apache.spark.storage.memory.MemoryStore.putIteratorAsBytes(MemoryStore.scala:349)\n",
      "\tat org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:1182)\n",
      "\tat org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:1156)\n",
      "\tat org.apache.spark.storage.BlockManager.doPut(BlockManager.scala:1091)\n",
      "\tat org.apache.spark.storage.BlockManager.doPutIterator(BlockManager.scala:1156)\n",
      "\tat org.apache.spark.storage.BlockManager.getOrElseUpdate(BlockManager.scala:882)\n",
      "\tat org.apache.spark.rdd.RDD.getOrCompute(RDD.scala:335)\n",
      "\tat org.apache.spark.rdd.RDD.iterator(RDD.scala:286)\n",
      "\tat org.apache.spark.api.python.PythonRDD.compute(PythonRDD.scala:65)\n",
      "\tat org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)\n",
      "\tat org.apache.spark.rdd.RDD.iterator(RDD.scala:288)\n",
      "\tat org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)\n",
      "\tat org.apache.spark.scheduler.Task.run(Task.scala:121)\n",
      "\tat org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)\n",
      "\tat org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)\n",
      "\tat org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)\n",
      "\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n",
      "\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n",
      "\tat java.lang.Thread.run(Thread.java:748)\n",
      "2021-02-02 20:27:33 ERROR Executor:91 - Exception in task 1.0 in stage 2.0 (TID 9)\n",
      "org.apache.spark.api.python.PythonException: Traceback (most recent call last):\n",
      "  File \"/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/worker.py\", line 377, in main\n",
      "    process()\n",
      "  File \"/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/worker.py\", line 372, in process\n",
      "    serializer.dump_stream(func(split_index, iterator), outfile)\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/data/ray_rdd.py\", line 133, in <lambda>\n",
      "    lambda idx, _: get_from_ray(idx, address, password, meta_store_name))\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/data/ray_rdd.py\", line 99, in get_from_ray\n",
      "    partition = ray.get(object_id)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/ray/worker.py\", line 1513, in get\n",
      "    raise value.as_instanceof_cause()\n",
      "ray.exceptions.RayTaskError(OverflowError): \u001b[36mray::TFRunner.predict()\u001b[39m (pid=18170, ip=10.239.44.107)\n",
      "  File \"python/ray/_raylet.pyx\", line 452, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 407, in ray._raylet.execute_task.function_executor\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in predict\n",
      "    new_part = [predict_fn(shard) for shard in partition]\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in <listcomp>\n",
      "    new_part = [predict_fn(shard) for shard in partition]\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 448, in predict_fn\n",
      "    y = local_model.predict(shard[\"x\"], **params)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 130, in _method_wrapper\n",
      "    return method(self, *args, **kwargs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 1613, in predict\n",
      "    callbacks.on_predict_end()\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 582, in on_predict_end\n",
      "    callback.on_predict_end(logs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 979, in on_predict_end\n",
      "    self._finalize_progbar(logs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 1026, in _finalize_progbar\n",
      "    self.progbar.update(self.seen, list(logs.items()), finalize=True)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py\", line 581, in update\n",
      "    numdigits = int(np.log10(self.target)) + 1\n",
      "OverflowError: cannot convert float infinity to integer\n",
      "\n",
      "\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:452)\n",
      "\tat org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:588)\n",
      "\tat org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:571)\n",
      "\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:406)\n",
      "\tat org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)\n",
      "\tat org.apache.spark.storage.memory.MemoryStore.putIterator(MemoryStore.scala:221)\n",
      "\tat org.apache.spark.storage.memory.MemoryStore.putIteratorAsBytes(MemoryStore.scala:349)\n",
      "\tat org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:1182)\n",
      "\tat org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:1156)\n",
      "\tat org.apache.spark.storage.BlockManager.doPut(BlockManager.scala:1091)\n",
      "\tat org.apache.spark.storage.BlockManager.doPutIterator(BlockManager.scala:1156)\n",
      "\tat org.apache.spark.storage.BlockManager.getOrElseUpdate(BlockManager.scala:882)\n",
      "\tat org.apache.spark.rdd.RDD.getOrCompute(RDD.scala:335)\n",
      "\tat org.apache.spark.rdd.RDD.iterator(RDD.scala:286)\n",
      "\tat org.apache.spark.api.python.PythonRDD.compute(PythonRDD.scala:65)\n",
      "\tat org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)\n",
      "\tat org.apache.spark.rdd.RDD.iterator(RDD.scala:288)\n",
      "\tat org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)\n",
      "\tat org.apache.spark.scheduler.Task.run(Task.scala:121)\n",
      "\tat org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)\n",
      "\tat org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)\n",
      "\tat org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)\n",
      "\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n",
      "\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n",
      "\tat java.lang.Thread.run(Thread.java:748)\n",
      "2021-02-02 20:27:33 ERROR TaskSetManager:70 - Task 1 in stage 2.0 failed 1 times; aborting job\n",
      "2021-02-02 20:27:33 ERROR TaskSetManager:70 - Task 1 in stage 2.0 failed 1 times; aborting job\n",
      "2021-02-02 20:27:33 ERROR TaskSetManager:70 - Task 1 in stage 2.0 failed 1 times; aborting job\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(pid=18170)\u001b[0m \r\n",
      "\u001b[2m\u001b[36m(pid=18170)\u001b[0m /home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py:581: RuntimeWarning: divide by zero encountered in log10\n",
      "\u001b[2m\u001b[36m(pid=18170)\u001b[0m   numdigits = int(np.log10(self.target)) + 1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(pid=18170)\u001b[0m \r\n",
      "\u001b[2m\u001b[36m(pid=18170)\u001b[0m /home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py:581: RuntimeWarning: divide by zero encountered in log10\n",
      "\u001b[2m\u001b[36m(pid=18170)\u001b[0m   numdigits = int(np.log10(self.target)) + 1\n",
      "\u001b[2m\u001b[36m(pid=18170)\u001b[0m \r\n",
      "\u001b[2m\u001b[36m(pid=18170)\u001b[0m /home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py:581: RuntimeWarning: divide by zero encountered in log10\n",
      "\u001b[2m\u001b[36m(pid=18170)\u001b[0m   numdigits = int(np.log10(self.target)) + 1\n"
     ]
    },
    {
     "ename": "Py4JJavaError",
     "evalue": "An error occurred while calling z:org.apache.spark.api.python.PythonRDD.collectAndServe.\n: org.apache.spark.SparkException: Job aborted due to stage failure: Task 1 in stage 2.0 failed 1 times, most recent failure: Lost task 1.0 in stage 2.0 (TID 9, localhost, executor driver): org.apache.spark.api.python.PythonException: Traceback (most recent call last):\n  File \"/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/worker.py\", line 377, in main\n    process()\n  File \"/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/worker.py\", line 372, in process\n    serializer.dump_stream(func(split_index, iterator), outfile)\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/data/ray_rdd.py\", line 133, in <lambda>\n    lambda idx, _: get_from_ray(idx, address, password, meta_store_name))\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/data/ray_rdd.py\", line 99, in get_from_ray\n    partition = ray.get(object_id)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/ray/worker.py\", line 1513, in get\n    raise value.as_instanceof_cause()\nray.exceptions.RayTaskError(OverflowError): \u001b[36mray::TFRunner.predict()\u001b[39m (pid=18170, ip=10.239.44.107)\n  File \"python/ray/_raylet.pyx\", line 452, in ray._raylet.execute_task\n  File \"python/ray/_raylet.pyx\", line 407, in ray._raylet.execute_task.function_executor\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in predict\n    new_part = [predict_fn(shard) for shard in partition]\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in <listcomp>\n    new_part = [predict_fn(shard) for shard in partition]\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 448, in predict_fn\n    y = local_model.predict(shard[\"x\"], **params)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 130, in _method_wrapper\n    return method(self, *args, **kwargs)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 1613, in predict\n    callbacks.on_predict_end()\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 582, in on_predict_end\n    callback.on_predict_end(logs)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 979, in on_predict_end\n    self._finalize_progbar(logs)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 1026, in _finalize_progbar\n    self.progbar.update(self.seen, list(logs.items()), finalize=True)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py\", line 581, in update\n    numdigits = int(np.log10(self.target)) + 1\nOverflowError: cannot convert float infinity to integer\n\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:452)\n\tat org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:588)\n\tat org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:571)\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:406)\n\tat org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)\n\tat org.apache.spark.storage.memory.MemoryStore.putIterator(MemoryStore.scala:221)\n\tat org.apache.spark.storage.memory.MemoryStore.putIteratorAsBytes(MemoryStore.scala:349)\n\tat org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:1182)\n\tat org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:1156)\n\tat org.apache.spark.storage.BlockManager.doPut(BlockManager.scala:1091)\n\tat org.apache.spark.storage.BlockManager.doPutIterator(BlockManager.scala:1156)\n\tat org.apache.spark.storage.BlockManager.getOrElseUpdate(BlockManager.scala:882)\n\tat org.apache.spark.rdd.RDD.getOrCompute(RDD.scala:335)\n\tat org.apache.spark.rdd.RDD.iterator(RDD.scala:286)\n\tat org.apache.spark.api.python.PythonRDD.compute(PythonRDD.scala:65)\n\tat org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)\n\tat org.apache.spark.rdd.RDD.iterator(RDD.scala:288)\n\tat org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)\n\tat org.apache.spark.scheduler.Task.run(Task.scala:121)\n\tat org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)\n\tat org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)\n\tat org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n\tat java.lang.Thread.run(Thread.java:748)\n\nDriver stacktrace:\n\tat org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1889)\n\tat org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1877)\n\tat org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1876)\n\tat scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)\n\tat scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)\n\tat org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1876)\n\tat org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:926)\n\tat org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:926)\n\tat scala.Option.foreach(Option.scala:257)\n\tat org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:926)\n\tat org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2110)\n\tat org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2059)\n\tat org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2048)\n\tat org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)\n\tat org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:737)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2061)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2082)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2101)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2126)\n\tat org.apache.spark.rdd.RDD$$anonfun$collect$1.apply(RDD.scala:945)\n\tat org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)\n\tat org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)\n\tat org.apache.spark.rdd.RDD.withScope(RDD.scala:363)\n\tat org.apache.spark.rdd.RDD.collect(RDD.scala:944)\n\tat org.apache.spark.api.python.PythonRDD$.collectAndServe(PythonRDD.scala:166)\n\tat org.apache.spark.api.python.PythonRDD.collectAndServe(PythonRDD.scala)\n\tat sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n\tat sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\n\tat sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n\tat java.lang.reflect.Method.invoke(Method.java:498)\n\tat py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)\n\tat py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)\n\tat py4j.Gateway.invoke(Gateway.java:282)\n\tat py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)\n\tat py4j.commands.CallCommand.execute(CallCommand.java:79)\n\tat py4j.GatewayConnection.run(GatewayConnection.java:238)\n\tat java.lang.Thread.run(Thread.java:748)\nCaused by: org.apache.spark.api.python.PythonException: Traceback (most recent call last):\n  File \"/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/worker.py\", line 377, in main\n    process()\n  File \"/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/worker.py\", line 372, in process\n    serializer.dump_stream(func(split_index, iterator), outfile)\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/data/ray_rdd.py\", line 133, in <lambda>\n    lambda idx, _: get_from_ray(idx, address, password, meta_store_name))\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/data/ray_rdd.py\", line 99, in get_from_ray\n    partition = ray.get(object_id)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/ray/worker.py\", line 1513, in get\n    raise value.as_instanceof_cause()\nray.exceptions.RayTaskError(OverflowError): \u001b[36mray::TFRunner.predict()\u001b[39m (pid=18170, ip=10.239.44.107)\n  File \"python/ray/_raylet.pyx\", line 452, in ray._raylet.execute_task\n  File \"python/ray/_raylet.pyx\", line 407, in ray._raylet.execute_task.function_executor\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in predict\n    new_part = [predict_fn(shard) for shard in partition]\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in <listcomp>\n    new_part = [predict_fn(shard) for shard in partition]\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 448, in predict_fn\n    y = local_model.predict(shard[\"x\"], **params)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 130, in _method_wrapper\n    return method(self, *args, **kwargs)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 1613, in predict\n    callbacks.on_predict_end()\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 582, in on_predict_end\n    callback.on_predict_end(logs)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 979, in on_predict_end\n    self._finalize_progbar(logs)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 1026, in _finalize_progbar\n    self.progbar.update(self.seen, list(logs.items()), finalize=True)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py\", line 581, in update\n    numdigits = int(np.log10(self.target)) + 1\nOverflowError: cannot convert float infinity to integer\n\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:452)\n\tat org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:588)\n\tat org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:571)\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:406)\n\tat org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)\n\tat org.apache.spark.storage.memory.MemoryStore.putIterator(MemoryStore.scala:221)\n\tat org.apache.spark.storage.memory.MemoryStore.putIteratorAsBytes(MemoryStore.scala:349)\n\tat org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:1182)\n\tat org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:1156)\n\tat org.apache.spark.storage.BlockManager.doPut(BlockManager.scala:1091)\n\tat org.apache.spark.storage.BlockManager.doPutIterator(BlockManager.scala:1156)\n\tat org.apache.spark.storage.BlockManager.getOrElseUpdate(BlockManager.scala:882)\n\tat org.apache.spark.rdd.RDD.getOrCompute(RDD.scala:335)\n\tat org.apache.spark.rdd.RDD.iterator(RDD.scala:286)\n\tat org.apache.spark.api.python.PythonRDD.compute(PythonRDD.scala:65)\n\tat org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)\n\tat org.apache.spark.rdd.RDD.iterator(RDD.scala:288)\n\tat org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)\n\tat org.apache.spark.scheduler.Task.run(Task.scala:121)\n\tat org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)\n\tat org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)\n\tat org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n\t... 1 more\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mPy4JJavaError\u001b[0m                             Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-18-78923351064b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mpreds\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mest\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpredict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msample_shards\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbatch_size\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcollect\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/analytics-zoo/pyzoo/zoo/orca/learn/tf2/estimator.py\u001b[0m in \u001b[0;36mpredict\u001b[0;34m(self, data, batch_size, verbose, steps, callbacks, data_config, feature_cols)\u001b[0m\n\u001b[1;32m    303\u001b[0m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mconvert_predict_xshards_to_dataframe\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpred_shards\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    304\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mSparkXShards\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 305\u001b[0;31m             \u001b[0mpred_shards\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_predict_spark_xshards\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mparams\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    306\u001b[0m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mupdate_predict_xshards\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpred_shards\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    307\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/analytics-zoo/pyzoo/zoo/orca/learn/tf2/estimator.py\u001b[0m in \u001b[0;36m_predict_spark_xshards\u001b[0;34m(self, xshards, params)\u001b[0m\n\u001b[1;32m    276\u001b[0m                                                                \u001b[0mtransform_func\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    277\u001b[0m                                                                gang_scheduling=False)\n\u001b[0;32m--> 278\u001b[0;31m         \u001b[0mspark_xshards\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpred_shards\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_spark_xshards\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    279\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mspark_xshards\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    280\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/analytics-zoo/pyzoo/zoo/orca/data/shard.py\u001b[0m in \u001b[0;36mto_spark_xshards\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    160\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    161\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mto_spark_xshards\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 162\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mSparkXShards\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mray_rdd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_spark_rdd\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    163\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    164\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mstaticmethod\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/analytics-zoo/pyzoo/zoo/orca/data/shard.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, rdd, transient)\u001b[0m\n\u001b[1;32m    182\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrdd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcache\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    183\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0meager\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 184\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    185\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtype\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    186\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/analytics-zoo/pyzoo/zoo/orca/data/shard.py\u001b[0m in \u001b[0;36mcompute\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    238\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    239\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mcompute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 240\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrdd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcount\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    241\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    242\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/rdd.py\u001b[0m in \u001b[0;36mcount\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1053\u001b[0m         \u001b[0;36m3\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1054\u001b[0m         \"\"\"\n\u001b[0;32m-> 1055\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmapPartitions\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0m_\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1056\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1057\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mstats\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/rdd.py\u001b[0m in \u001b[0;36msum\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1044\u001b[0m         \u001b[0;36m6.0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1045\u001b[0m         \"\"\"\n\u001b[0;32m-> 1046\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmapPartitions\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfold\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moperator\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madd\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1047\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1048\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mcount\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/rdd.py\u001b[0m in \u001b[0;36mfold\u001b[0;34m(self, zeroValue, op)\u001b[0m\n\u001b[1;32m    915\u001b[0m         \u001b[0;31m# zeroValue provided to each partition is unique from the one provided\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    916\u001b[0m         \u001b[0;31m# to the final reduce call\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 917\u001b[0;31m         \u001b[0mvals\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmapPartitions\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfunc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcollect\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    918\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mreduce\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mop\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvals\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mzeroValue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    919\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/rdd.py\u001b[0m in \u001b[0;36mcollect\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    814\u001b[0m         \"\"\"\n\u001b[1;32m    815\u001b[0m         \u001b[0;32mwith\u001b[0m \u001b[0mSCCallSiteSync\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcontext\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mcss\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 816\u001b[0;31m             \u001b[0msock_info\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mctx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jvm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mPythonRDD\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcollectAndServe\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jrdd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrdd\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    817\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_load_from_socket\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msock_info\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_jrdd_deserializer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    818\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/py4j-0.10.7-src.zip/py4j/java_gateway.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args)\u001b[0m\n\u001b[1;32m   1255\u001b[0m         \u001b[0manswer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgateway_client\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend_command\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcommand\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1256\u001b[0m         return_value = get_return_value(\n\u001b[0;32m-> 1257\u001b[0;31m             answer, self.gateway_client, self.target_id, self.name)\n\u001b[0m\u001b[1;32m   1258\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1259\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mtemp_arg\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mtemp_args\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/py4j-0.10.7-src.zip/py4j/protocol.py\u001b[0m in \u001b[0;36mget_return_value\u001b[0;34m(answer, gateway_client, target_id, name)\u001b[0m\n\u001b[1;32m    326\u001b[0m                 raise Py4JJavaError(\n\u001b[1;32m    327\u001b[0m                     \u001b[0;34m\"An error occurred while calling {0}{1}{2}.\\n\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 328\u001b[0;31m                     format(target_id, \".\", name), value)\n\u001b[0m\u001b[1;32m    329\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    330\u001b[0m                 raise Py4JError(\n",
      "\u001b[0;31mPy4JJavaError\u001b[0m: An error occurred while calling z:org.apache.spark.api.python.PythonRDD.collectAndServe.\n: org.apache.spark.SparkException: Job aborted due to stage failure: Task 1 in stage 2.0 failed 1 times, most recent failure: Lost task 1.0 in stage 2.0 (TID 9, localhost, executor driver): org.apache.spark.api.python.PythonException: Traceback (most recent call last):\n  File \"/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/worker.py\", line 377, in main\n    process()\n  File \"/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/worker.py\", line 372, in process\n    serializer.dump_stream(func(split_index, iterator), outfile)\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/data/ray_rdd.py\", line 133, in <lambda>\n    lambda idx, _: get_from_ray(idx, address, password, meta_store_name))\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/data/ray_rdd.py\", line 99, in get_from_ray\n    partition = ray.get(object_id)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/ray/worker.py\", line 1513, in get\n    raise value.as_instanceof_cause()\nray.exceptions.RayTaskError(OverflowError): \u001b[36mray::TFRunner.predict()\u001b[39m (pid=18170, ip=10.239.44.107)\n  File \"python/ray/_raylet.pyx\", line 452, in ray._raylet.execute_task\n  File \"python/ray/_raylet.pyx\", line 407, in ray._raylet.execute_task.function_executor\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in predict\n    new_part = [predict_fn(shard) for shard in partition]\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in <listcomp>\n    new_part = [predict_fn(shard) for shard in partition]\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 448, in predict_fn\n    y = local_model.predict(shard[\"x\"], **params)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 130, in _method_wrapper\n    return method(self, *args, **kwargs)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 1613, in predict\n    callbacks.on_predict_end()\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 582, in on_predict_end\n    callback.on_predict_end(logs)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 979, in on_predict_end\n    self._finalize_progbar(logs)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 1026, in _finalize_progbar\n    self.progbar.update(self.seen, list(logs.items()), finalize=True)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py\", line 581, in update\n    numdigits = int(np.log10(self.target)) + 1\nOverflowError: cannot convert float infinity to integer\n\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:452)\n\tat org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:588)\n\tat org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:571)\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:406)\n\tat org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)\n\tat org.apache.spark.storage.memory.MemoryStore.putIterator(MemoryStore.scala:221)\n\tat org.apache.spark.storage.memory.MemoryStore.putIteratorAsBytes(MemoryStore.scala:349)\n\tat org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:1182)\n\tat org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:1156)\n\tat org.apache.spark.storage.BlockManager.doPut(BlockManager.scala:1091)\n\tat org.apache.spark.storage.BlockManager.doPutIterator(BlockManager.scala:1156)\n\tat org.apache.spark.storage.BlockManager.getOrElseUpdate(BlockManager.scala:882)\n\tat org.apache.spark.rdd.RDD.getOrCompute(RDD.scala:335)\n\tat org.apache.spark.rdd.RDD.iterator(RDD.scala:286)\n\tat org.apache.spark.api.python.PythonRDD.compute(PythonRDD.scala:65)\n\tat org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)\n\tat org.apache.spark.rdd.RDD.iterator(RDD.scala:288)\n\tat org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)\n\tat org.apache.spark.scheduler.Task.run(Task.scala:121)\n\tat org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)\n\tat org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)\n\tat org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n\tat java.lang.Thread.run(Thread.java:748)\n\nDriver stacktrace:\n\tat org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1889)\n\tat org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1877)\n\tat org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1876)\n\tat scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)\n\tat scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)\n\tat org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1876)\n\tat org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:926)\n\tat org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:926)\n\tat scala.Option.foreach(Option.scala:257)\n\tat org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:926)\n\tat org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2110)\n\tat org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2059)\n\tat org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2048)\n\tat org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)\n\tat org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:737)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2061)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2082)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2101)\n\tat org.apache.spark.SparkContext.runJob(SparkContext.scala:2126)\n\tat org.apache.spark.rdd.RDD$$anonfun$collect$1.apply(RDD.scala:945)\n\tat org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)\n\tat org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)\n\tat org.apache.spark.rdd.RDD.withScope(RDD.scala:363)\n\tat org.apache.spark.rdd.RDD.collect(RDD.scala:944)\n\tat org.apache.spark.api.python.PythonRDD$.collectAndServe(PythonRDD.scala:166)\n\tat org.apache.spark.api.python.PythonRDD.collectAndServe(PythonRDD.scala)\n\tat sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n\tat sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\n\tat sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n\tat java.lang.reflect.Method.invoke(Method.java:498)\n\tat py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)\n\tat py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)\n\tat py4j.Gateway.invoke(Gateway.java:282)\n\tat py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)\n\tat py4j.commands.CallCommand.execute(CallCommand.java:79)\n\tat py4j.GatewayConnection.run(GatewayConnection.java:238)\n\tat java.lang.Thread.run(Thread.java:748)\nCaused by: org.apache.spark.api.python.PythonException: Traceback (most recent call last):\n  File \"/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/worker.py\", line 377, in main\n    process()\n  File \"/intern/spark/spark-2.4.3-bin-hadoop2.7/python/lib/pyspark.zip/pyspark/worker.py\", line 372, in process\n    serializer.dump_stream(func(split_index, iterator), outfile)\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/data/ray_rdd.py\", line 133, in <lambda>\n    lambda idx, _: get_from_ray(idx, address, password, meta_store_name))\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/data/ray_rdd.py\", line 99, in get_from_ray\n    partition = ray.get(object_id)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/ray/worker.py\", line 1513, in get\n    raise value.as_instanceof_cause()\nray.exceptions.RayTaskError(OverflowError): \u001b[36mray::TFRunner.predict()\u001b[39m (pid=18170, ip=10.239.44.107)\n  File \"python/ray/_raylet.pyx\", line 452, in ray._raylet.execute_task\n  File \"python/ray/_raylet.pyx\", line 407, in ray._raylet.execute_task.function_executor\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in predict\n    new_part = [predict_fn(shard) for shard in partition]\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in <listcomp>\n    new_part = [predict_fn(shard) for shard in partition]\n  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 448, in predict_fn\n    y = local_model.predict(shard[\"x\"], **params)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 130, in _method_wrapper\n    return method(self, *args, **kwargs)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 1613, in predict\n    callbacks.on_predict_end()\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 582, in on_predict_end\n    callback.on_predict_end(logs)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 979, in on_predict_end\n    self._finalize_progbar(logs)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 1026, in _finalize_progbar\n    self.progbar.update(self.seen, list(logs.items()), finalize=True)\n  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py\", line 581, in update\n    numdigits = int(np.log10(self.target)) + 1\nOverflowError: cannot convert float infinity to integer\n\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.handlePythonException(PythonRunner.scala:452)\n\tat org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:588)\n\tat org.apache.spark.api.python.PythonRunner$$anon$1.read(PythonRunner.scala:571)\n\tat org.apache.spark.api.python.BasePythonRunner$ReaderIterator.hasNext(PythonRunner.scala:406)\n\tat org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)\n\tat org.apache.spark.storage.memory.MemoryStore.putIterator(MemoryStore.scala:221)\n\tat org.apache.spark.storage.memory.MemoryStore.putIteratorAsBytes(MemoryStore.scala:349)\n\tat org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:1182)\n\tat org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:1156)\n\tat org.apache.spark.storage.BlockManager.doPut(BlockManager.scala:1091)\n\tat org.apache.spark.storage.BlockManager.doPutIterator(BlockManager.scala:1156)\n\tat org.apache.spark.storage.BlockManager.getOrElseUpdate(BlockManager.scala:882)\n\tat org.apache.spark.rdd.RDD.getOrCompute(RDD.scala:335)\n\tat org.apache.spark.rdd.RDD.iterator(RDD.scala:286)\n\tat org.apache.spark.api.python.PythonRDD.compute(PythonRDD.scala:65)\n\tat org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)\n\tat org.apache.spark.rdd.RDD.iterator(RDD.scala:288)\n\tat org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)\n\tat org.apache.spark.scheduler.Task.run(Task.scala:121)\n\tat org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)\n\tat org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)\n\tat org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)\n\tat java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n\t... 1 more\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 2:==============>                                            (1 + 2) / 4]\u001b[2m\u001b[36m(pid=18170)\u001b[0m \n",
      "[Stage 2:==============>                                            (1 + 1) / 4]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(pid=18170)\u001b[0m \n",
      "\u001b[2m\u001b[36m(pid=18170)\u001b[0m \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-02-02 20:27:38,505\tERROR worker.py:1012 -- Possible unhandled error from worker: \u001b[36mray::TFRunner.predict()\u001b[39m (pid=18170, ip=10.239.44.107)\n",
      "  File \"python/ray/_raylet.pyx\", line 452, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 407, in ray._raylet.execute_task.function_executor\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in predict\n",
      "    new_part = [predict_fn(shard) for shard in partition]\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in <listcomp>\n",
      "    new_part = [predict_fn(shard) for shard in partition]\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 448, in predict_fn\n",
      "    y = local_model.predict(shard[\"x\"], **params)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 130, in _method_wrapper\n",
      "    return method(self, *args, **kwargs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 1613, in predict\n",
      "    callbacks.on_predict_end()\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 582, in on_predict_end\n",
      "    callback.on_predict_end(logs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 979, in on_predict_end\n",
      "    self._finalize_progbar(logs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 1026, in _finalize_progbar\n",
      "    self.progbar.update(self.seen, list(logs.items()), finalize=True)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py\", line 581, in update\n",
      "    numdigits = int(np.log10(self.target)) + 1\n",
      "OverflowError: cannot convert float infinity to integer\n",
      "2021-02-02 20:27:39,506\tERROR worker.py:1012 -- Possible unhandled error from worker: \u001b[36mray::TFRunner.predict()\u001b[39m (pid=18170, ip=10.239.44.107)\n",
      "  File \"python/ray/_raylet.pyx\", line 452, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 407, in ray._raylet.execute_task.function_executor\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in predict\n",
      "    new_part = [predict_fn(shard) for shard in partition]\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in <listcomp>\n",
      "    new_part = [predict_fn(shard) for shard in partition]\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 448, in predict_fn\n",
      "    y = local_model.predict(shard[\"x\"], **params)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 130, in _method_wrapper\n",
      "    return method(self, *args, **kwargs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 1613, in predict\n",
      "    callbacks.on_predict_end()\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 582, in on_predict_end\n",
      "    callback.on_predict_end(logs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 979, in on_predict_end\n",
      "    self._finalize_progbar(logs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 1026, in _finalize_progbar\n",
      "    self.progbar.update(self.seen, list(logs.items()), finalize=True)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py\", line 581, in update\n",
      "    numdigits = int(np.log10(self.target)) + 1\n",
      "OverflowError: cannot convert float infinity to integer\n",
      "2021-02-02 20:27:39,506\tERROR worker.py:1012 -- Possible unhandled error from worker: \u001b[36mray::TFRunner.predict()\u001b[39m (pid=18170, ip=10.239.44.107)\n",
      "  File \"python/ray/_raylet.pyx\", line 452, in ray._raylet.execute_task\n",
      "  File \"python/ray/_raylet.pyx\", line 407, in ray._raylet.execute_task.function_executor\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in predict\n",
      "    new_part = [predict_fn(shard) for shard in partition]\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 451, in <listcomp>\n",
      "    new_part = [predict_fn(shard) for shard in partition]\n",
      "  File \"/home/zhenhao/analytics-zoo/pyzoo/zoo/orca/learn/tf2/tf_runner.py\", line 448, in predict_fn\n",
      "    y = local_model.predict(shard[\"x\"], **params)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 130, in _method_wrapper\n",
      "    return method(self, *args, **kwargs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py\", line 1613, in predict\n",
      "    callbacks.on_predict_end()\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 582, in on_predict_end\n",
      "    callback.on_predict_end(logs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 979, in on_predict_end\n",
      "    self._finalize_progbar(logs)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/callbacks.py\", line 1026, in _finalize_progbar\n",
      "    self.progbar.update(self.seen, list(logs.items()), finalize=True)\n",
      "  File \"/home/zhenhao/.local/lib/python3.7/site-packages/tensorflow/python/keras/utils/generic_utils.py\", line 581, in update\n",
      "    numdigits = int(np.log10(self.target)) + 1\n",
      "OverflowError: cannot convert float infinity to integer\n"
     ]
    }
   ],
   "source": [
    "preds = est.predict(sample_shards, batch_size=1).collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "golden-treasurer",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,\n",
       "       1., 1., 1., 1., 1., 1., 1., 1., 1.])"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampled[0].sum(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "third-drunk",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
